File: paimon-python/pypaimon/tests/global_index_schema_compatibility_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import types
import unittest

from pypaimon.globalindex.global_index_meta import GlobalIndexMeta
from pypaimon.globalindex.global_index_schema_compatibility import (
    filter_compatible_global_indexes,
)
from pypaimon.index.index_file_meta import IndexFileMeta
from pypaimon.manifest.index_manifest_entry import IndexManifestEntry
from pypaimon.schema.data_types import ArrayType, AtomicType, DataField
from pypaimon.schema.table_schema import TableSchema


def _field(field_id, name, data_type):
    return DataField(field_id, name, data_type)


def _schema(schema_id, fields):
    return TableSchema(id=schema_id, fields=fields)


def _entry(file_name, schema_id, field_id, extra_field_ids=None):
    index_file = IndexFileMeta(
        index_type='bitmap',
        file_name=file_name,
        file_size=1,
        row_count=1,
        global_index_meta=GlobalIndexMeta(
            row_range_start=0,
            row_range_end=0,
            index_field_id=field_id,
            extra_field_ids=extra_field_ids,
        ),
    )
    return IndexManifestEntry(0, None, 0, index_file, schema_id)


class GlobalIndexSchemaCompatibilityTest(unittest.TestCase):

    def test_filters_by_indexed_field_types(self):
        historical = _schema(1, [
            _field(1, 'numbers', ArrayType(True, AtomicType('INT'))),
            _field(2, 'name', AtomicType('STRING', nullable=False)),
            _field(3, 'age', AtomicType('INT')),
        ])
        current = _schema(2, [
            _field(1, 'numbers', ArrayType(True, AtomicType('BIGINT'))),
            _field(2, 'renamed', AtomicType('STRING')),
            _field(3, 'age', AtomicType('BIGINT')),
        ])
        schema_lookups = []

        def get_schema(schema_id):
            schema_lookups.append(schema_id)
            return historical if schema_id == historical.id else None

        table = types.SimpleNamespace(
            table_schema=current,
            schema_manager=types.SimpleNamespace(get_schema=get_schema),
        )
        compatible = _entry('compatible', 1, 2)
        current_schema = _entry('current', 2, 2)
        changed_primary = _entry('changed-primary', 1, 1)
        changed_extra = _entry('changed-extra', 1, 2, [3])
        legacy = _entry('legacy', None, 2)
        missing_schema_a = _entry('missing-a', 99, 2)
        missing_schema_b = _entry('missing-b', 99, 2)

        result = filter_compatible_global_indexes(table, [
            compatible,
            current_schema,
            changed_primary,
            changed_extra,
            legacy,
            missing_schema_a,
            missing_schema_b,
        ])

        self.assertEqual(
            ['compatible', 'current'],
            [entry.index_file.file_name for entry in result],
        )
        self.assertEqual([1, 99], schema_lookups)


if __name__ == '__main__':
    unittest.main()

File: paimon-core/src/test/java/org/apache/paimon/globalindex/GlobalIndexSchemaCompatibilityTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.globalindex;

import org.apache.paimon.CoreOptions;
import org.apache.paimon.data.BinaryRow;
import org.apache.paimon.index.GlobalIndexMeta;
import org.apache.paimon.index.IndexFileMeta;
import org.apache.paimon.manifest.FileKind;
import org.apache.paimon.manifest.IndexManifestEntry;
import org.apache.paimon.schema.Schema;
import org.apache.paimon.schema.SchemaChange;
import org.apache.paimon.table.FileStoreTable;
import org.apache.paimon.table.TableTestBase;
import org.apache.paimon.types.DataTypes;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.assertj.core.api.Assertions.assertThat;

/** Tests for {@link GlobalIndexSchemaCompatibility}. */
public class GlobalIndexSchemaCompatibilityTest extends TableTestBase {

    @Override
    protected Schema schemaDefault() {
        return Schema.newBuilder()
                .column("indexed", DataTypes.INT())
                .option(CoreOptions.BUCKET.key(), "-1")
                .option(CoreOptions.ROW_TRACKING_ENABLED.key(), "true")
                .option(CoreOptions.DATA_EVOLUTION_ENABLED.key(), "true")
                .build();
    }

    @Test
    public void testCompatibilityUsesIndexedFieldTypes() throws Exception {
        createTableDefault();
        FileStoreTable table = getTableDefault();
        long buildSchemaId = table.schema().id();
        IndexManifestEntry entry = globalIndexEntry(buildSchemaId);

        catalog.alterTable(
                identifier(),
                Collections.singletonList(SchemaChange.addColumn("unrelated", DataTypes.STRING())),
                false);
        table = getTableDefault();
        assertThat(
                        GlobalIndexSchemaCompatibility.filterCompatible(
                                table, Collections.singleton(entry)))
                .containsExactly(entry);

        catalog.alterTable(
                identifier(),
                Collections.singletonList(
                        SchemaChange.updateColumnType("indexed", DataTypes.BIGINT())),
                false);
        table = getTableDefault();
        assertThat(
                        GlobalIndexSchemaCompatibility.filterCompatible(
                                table, Collections.singleton(entry)))
                .isEmpty();
    }

    @Test
    public void testMissingSchemaIdentityFailsClosed() throws Exception {
        createTableDefault();
        FileStoreTable table = getTableDefault();

        assertThat(
                        GlobalIndexSchemaCompatibility.filterCompatible(
                                table,
                                Arrays.asList(
                                        globalIndexEntry(null), globalIndexEntry(Long.MAX_VALUE))))
                .isEmpty();
    }

    private static IndexManifestEntry globalIndexEntry(Long schemaId) {
        IndexFileMeta indexFile =
                new IndexFileMeta(
                        "btree",
                        "index-file",
                        1L,
                        1L,
                        null,
                        null,
                        new GlobalIndexMeta(0L, 0L, 0, null, null),
                        schemaId);
        return new IndexManifestEntry(FileKind.ADD, BinaryRow.EMPTY_ROW, 0, indexFile, schemaId);
    }
}
